Prevent JIT bodies from strictly outliving methods inlined into them

jdmpapin opened the pull request:
This PR will need a coordinated merge with eclipse-omr/omr#7673.

This is largely in preparation for constant references, but it should also have benefits for code motion. For more detail, see the Motivation section of the doc comment for `OMR::RetainedMethodSet`.

Much of this commit is `J9::RetainedMethodSet`, the OpenJ9 implementation of `OMR::RetainedMethodSet`, which describes a set of methods that will remain loaded under certain assumptions. It's represented as a set of class loaders together with a set of anonymous classes, since most classes are unloaded at class loader granularity, but anonymous classes can be unloaded individually.

Each `J9ClassLoader` now maintains a set of class loaders known to live at least as long at it (`outlivingLoaders`). When a class is looked up in a loader *L1*, and the class that's found was defined by a different loader *L2*, *L1* adds the class to its hash table, guaranteeing that *L2* will outlive *L1*. Now when this occurs, *L1* will additionally take note that *L2* outlives it. In this way we construct a graph of loaders describing the lifetime relationships that have been observed so far. This graph can be searched to determine the full set of loaders that are known to outlive any particular loader. It would be possible to skip reifying this graph and to get the same result instead by scanning through the (potentially much larger) class hash tables.

Given a particular method to assume, `J9::RetainedMethodSet` uses the graph of loaders to determine which other loaders can be inferred to remain loaded at least as long as the assumed method.

During inlining, each `TR_CallSite` and `TR_CallTarget` will now have an associated `RetainedMethodSet`, which may be used to arrange to establish certain lifetime relationships between the JIT body and inlined methods. It can extend the lifetime of an inlined method (keepalive) or restrict the lifetime of the JIT body (bond) to make sure that it won't run once an inlined method has been unloaded. See the corresponding OMR commit for more.

Keepalives are determined but not yet actually put into effect since we don't yet have any way to create the necessary references. Again see the OMR commit for more.

A bond is put into effect by creating an unload assumption so that when the inlined method is unloaded, the JIT body will be invalidated in much the same way as for preexistence. Once a JIT body has been invalidated for this reason, any later recompilation of the same method will set the JIT option `dontInlineUnloadableMethods`, which will cause the compiler to refuse to inline any target that would require a bond. This way, invalidation due to unloading can't cause more than one recompilation of a given method.

In AOT compilations, the compile-time analysis is skipped because its results can't be expected to hold at load time. Instead the analysis is done and runtime assumptions are created as necessary during AOT load, after all of the usual validations and relocations have succeeded, based on the inlining table in the `J9JITExceptionTable`. There are no missing entries because loads can only succeed if all inlined methods are found.

In addition to the keepalives mentioned above, the compilation object has a separate set of keepalive classes to be used for cases in which non-inlining transformations cause the IL to directly mention classes (or data belonging to classes) that might not be guaranteed to exist for the entire lifetime of the JIT body. This is used when refining `linkTo*` and `invokeBasic` calls in the trees and when `transformIndirectLoadChain()` transmutes a node into a class pointer `loadaddr`.

jdmpapin commented:
@0xdaryl and @vijaysun-omr, could you please review this together with eclipse-omr/omr#7673?
@dsouzai, could you please review the AOT part?
@mpirvu, could you please review the JITServer part?
@tajila, could you please review the VM part?

jdmpapin commented:
> The GC/stackwalker keeps inlined method classes alive - why is this necessary?

It only does so for JIT bodies currently on the stack (and I will be relying on that later). A JIT body that is *not* currently on the stack, however, can have its inlined methods unloaded. This change is to make sure that we won't re-enter the JIT body after that happens. Historically we have allowed the JIT body to be re-entered and we have tried to make the generated code cope with that situation. For example, we create runtime assumptions to patch profiled guards on unload, and loop versioner avoids versioning `checkcast` when the cast type might get unloaded before the outermost method.

My main motivation at the moment is to prepare for constant references (#16616). The compiler will create additional references that belong to the JIT body. If the JIT body is allowed to continue to run for the entire lifetime of the defining class of the outermost method, then these additional references need to persist for that long, and a combination of inlining and constant folding could cause a memory leak. If you want more detail, please see the GC and Unloading section of #16616 and the treatment of constant references in the Motivation section of the doc comment of `OMR::RetainedMethodSet` (in OMRRetainedMethodSet.hpp, eclipse-omr/omr#7673). I'll note here though that this is where I'll be relying on the stack walker. For any JIT body currently on the stack, the stack walker will prevent unloading of all (defining classes of) methods inlined into it, which will ensure that const refs for the JIT body will be considered live (if it has any). That will be important because the JIT body is still running, so it still might use its const refs in the future.

This will also benefit loop versioner. At the very least, we'll be able to allow it to always version `checkcast`, but also I'm almost certain that it currently does other transformations that are incorrect when the JIT body continues to run after inlined methods are unloaded.

gacholio commented:
> I'll note here though that this is where I'll be relying on the stack walker. For any JIT body currently on the stack, the stack walker prevents unloading of all (defining classes of) methods inlined into it.

The stack walker prevents unloading of inlined method classes currently running on the stack, not all classes that are inlined into a compiled method, I believe. The work is in `markClassesInInlineRanges`

jdmpapin commented:
`markClassesInInlineRanges()` looks to me like it walks the class of every inlined method. It [loops over all of the inlined methods](https://github.com/eclipse-openj9/openj9/blob/2dd07ebbd773e0b3e9243275ab271029a5959a57/runtime/compiler/runtime/MethodMetaData.c#L624-L642) from 0 to `getNumInlinedCallSites(methodMetaData)`. It only skips ones that satisfy `isPatchedValue()`, which IIUC are the ones that have already been unloaded (before this frame was entered). That's another measure that we currently have to allow the JIT body to be re-entered after an inlined method is unloaded, since if we left the inlining table unmodified, the GC would dereference dangling `J9Method` pointers.

That lines up with what I had observed experimentally when I wrote in #16616:
> Note that (as far as I have been able to tell) a JIT body on the stack already prevents unloading of all of the loaders of the methods inlined into it.

Following that, I argued that if I was mistaken (and so if I'm somehow mistaken right now), then it would need to be changed to work the way that I think it does:
> And just in case I'm mistaken, it *should* already do that even though we don't yet have constant references. It's important for the correctness of the generated code in cases where the compiler has been able to propagate a constant class pointer forward out of an inlined body and into the surrounding code, since that surrounding code might still be running.

gacholio commented on runtime/vm/KeyHashTable.c:
```diff
@@ -851,3 +851,58 @@ findClassLocationForClass(J9VMThread *currentThread, J9Class *clazz)
+				}
+			}
+		} else {
+			hashTableAdd((J9HashTable *)classLoader->outlivingLoaders, &outlivingLoader);
```
Is this an error (adding the address of a local into the hash table)?

gacholio replied:
Or is it going to be copied into the node by the API?

jdmpapin replied:
I believe it's copied. `&outlivingLoader` is a `J9ClassLoader**`, but the element type is `J9ClassLoader*` (only one level). If somehow this put the local's address into the hash table instead, then the JIT side of my change would surely crash, but I've seen it get the correct results based on the `J9ClassLoader` pointers it got from the hash table

jdmpapin replied:
Grepping for `hashTableAdd()`, it seems to be common to pass the address of a local variable.

jdmpapin commented:
There are two mechanisms to prevent unloading:
- Keepalive: A call site was refined based on known objects and so we consider ourselves within our rights to explicitly prevent unloading. This pair of PRs implements the analysis to determine which methods should get a keepalive, but it doesn't actually enact the keepalive. It will be implemented as an extra constant reference (along with the GC changes necessary to support those) later on. In the meantime, we assume that it won't be unloaded because we have a pervasive assumption (that I'm trying to eliminate) that when the compiler has determined that something is a known object, it won't be modified (or at least, once it is modified, we'll never reach the program point where the compiler made that determination).
- Bond: The compiler couldn't establish based on the class loader graph that the inlined method will remain loaded (even taking into account a potential keepalive for the call site). I'm specifically trying to avoid extending the lifetime of the inlined method to match that of the JIT body in this case. Otherwise, we'd get the kind of memory leak that I mentioned before, and without even doing any constant folding. Instead, we create a runtime assumption to invalidate the JIT body when the class that defines the inlined method is unloaded. At time of unloading, the JIT body is not on the stack, so invalidating it ensures that (beyond the preprologue) its code will never run again.

jdmpapin commented:
> Keepalive: [...] In the meantime, we assume that it won't be unloaded because...

I just realized that I could - maybe should - instead treat keepalives as extra bonds in the meantime, i.e. create unload assumptions for those as well. Then if our pervasive assumption fails to hold, it will just invalidate the JIT body instead of accidentally allowing it to be entered after inlined code has been unloaded.

mpirvu reviewed (commented):
LGTM. I wonder if there is a possibility of optimization from the JITServer point of view. There are 4 new messages an I see the following counts:
```
#0238    5661           RetainedMethodSet_createMirror
#0239     245           RetainedMethodSet_scan
#0240     164           RetainedMethodSet_keepalive
#0241     339           RetainedMethodSet_bond
```
Ideally we would avoid the `createMirror` message since it is the most frequent. As far I can tell, it's the `scan` message that is problematic for the server and needs client help. The `scan` needs the mirror, while the `keepalive` and `bond` are just to keep the mirror in sync.
The mirror is needed during the `scan` operation for `willRemainLoaded(outlivingLoader)` which traverses a hierarchy of RetainedMethodSets to find a match for the `outlivingLoader` passed as parameter.
What if the server combines the sets of `_loaders` for the entire `RetainedMethodSet` hierarchy and sends this as part of the `scan` message? Then, we would not need the mirrors at all at the client.

jdmpapin commented on runtime/compiler/compile/J9Compilation.cpp:
```diff
@@ -1671,6 +1677,59 @@ J9::Compilation::populateAOTMethodDependencies(TR_OpaqueClassBlock *definingClas
+      return false;
+      }
+
+   if (TR::Options::getCmdLineOptions()->getOption(TR_NoClassGC))
```
Oh, I think I did this just because I only found one place where we currently query `TR_NoClassGC`, which uses `TR::Options::getCmdLineOptions()`, and based on the way it gets set, it didn't seem to matter which `Options` we look at. But I guess on the server, `TR::Options::getCmdLineOptions()` will be for the server process? vs. the compilation's own options, which would reflect the situation at the client.

mpirvu replied:
Correct, `TR::Options::getCmdLineOptions()` uses the global options for the server and does not reflect the options set by the client.

jdmpapin replied:
Corrected in ef06f0c12efd51df6a389079baf1d3499a15e2b4

jdmpapin commented:
> What if the server combines the sets of `_loaders` for the entire `RetainedMethodSet` hierarchy and sends this as part of the `scan` message? Then, we would not need the mirrors at all at the client.

Every (or at least nearly every) new `RetainedMethodSet` entails a scan as part of its initialization. I combined the initial scan into the response for the `createMirror` message so that initializing a new `RetainedMethodSet` would only take one round trip instead of two. If from here we eliminated the mirrors, we would still scan, and the `createMirror` messages would all (or nearly all) just be replaced with `scan` messages. So we'd have essentially the same total number of messages, but we'd transfer more data because we'd have to keep sending the contents of all of the sets every time.

Here are the best ideas I can think of at the moment to reduce the number of messages:
1. Remove the `keepalive` and `bond` messages, and instead queue up keepalives and bonds to be sent along with the next `createMirror` or `scan` message (or with the `CHTableCommitData` if there is no later `createMirror` or `scan` message), since the keepalives and bonds won't be relevant on the client before that point anyway.
2. I might be able to create the compilation's root `RetainedMethodSet` less often. Outside of inlining, it looks like most of the uses could be restricted to the case where the set already exists, e.g. checking whether any bonds exist. There's one use where we're just trying to avoid adding a redundant keepalive, but we could just add the keepalive anyway. Within inlining, I might be able to play some tricks with `TR_CallSite`, but `TR_CallTarget` needs to know whether it `_needsBond`. Do you happen to know how commonly we compile leaf methods?
3. This idea is more involved. Outside of `dontInlineUnloadableMethods` (or even with that enabled, if we're OK restricting inlining more than necessary in that mode), we could entirely skip scanning during compilation on the server. The sets would just consist of the permanent loaders and whatever specific loaders and anonymous classes we've seen on a given call path. When we inline a `TR_CallSite` that wants a keepalive, we remember the method to keepalive and associate it with the inlined site index. Finally, at the end of the compilation, the server sends the client all of the remembered keepalive methods for the whole inlining table. At the moment I think this would be part of the `CHTableCommitData`, but later with const refs, if there appear to be keepalives, then I think it would be sent during codegen when we need to figure out which keepalives are actually needed. (Or we could potentially have the server skip trying to eliminate redundant keepalives, and just use all of them.) Then the client makes sure to set up `comp` to know about all of the inlined sites (if it doesn't already). Finally, the client repeats the analysis like in the AOT case, but this time doing all of the attestations, and taking into account keepalives whenever they aren't redundant for the inline call path. Essentially, we have the server do a less precise analysis, recording the information necessary for the client to do the proper analysis later, and send that to the client when we were going to be doing a round trip anyway.

mpirvu commented:
I would like to avoid complicating the code in exchange for a minuscule reduction in number of messages. If we cannot reduce the number of mirrors significantly (because a mirror means an implicit scan), then we should live the code as is.
I am wondering though how often we hit a "difficult" scan to be performed. The classloader introduces a new field, `outlivingLoaders`, which can be NULL, a singleton (identified by the `J9CLASSLOADER_OUTLIVING_LOADERS_SINGLE_TAG` tag) or a hashtable of classloaders. It's  the last case that the server cannot deal with. I would have to experiment to determine the frequency of each case and if the third case (the hashtable) occurs infrequently, we may have a shot at reducing the number of messages. This can be done as a separate PR, I think.

mpirvu commented:
For a given j9class the server knows the classloader of that class (as a pointer valid at the client). It cannot dereference the classloader, but it can cache the value of `outlivingLoaders`. Now, I realize there could be a problem because  `outlivingLoaders` can change at runtime (I think).

jdmpapin commented:
Yes, it changes as classes are looked up in initiating loaders at runtime

jdmpapin commented:
> I just realized that I could - maybe should - instead treat keepalives as extra bonds in the meantime, i.e. create unload assumptions for those as well.

I think it's probably better to leave this alone, since unexpected invalidations could be a problem of their own. In particular, `dontInlineUnloadableMethods` is supposed to guarantee that the resulting JIT body won't be invalidated by an unload assumption. This could allow a single method to be repeatedly invalidated and recompiled, but more importantly, we could end up trying to invalidate JIT bodies that don't support invalidation, e.g. `newInstance` thunks.

I mean for the implementation of constant references to the be the first later change that really makes essential use of this PR anyway, and keepalives will be fully implemented at the same time.

jdmpapin commented:
Updated to allow keepalive to be used for call sites created from call nodes that have been refined in the trees. This way, for inlining purposes, a top-level call refined by method handle transformer can keepalive (if needed) in exactly the same way as a call refined by `InterpreterEmulator`.

jdmpapin commented:
Moved the call to `keepaliveRefinedCallee()`  to flag nodes just after the main refinement transformation rather than just beforehand. Otherwise, refinement of `linkToVirtual()` would clear the flag. I don't think this move makes a difference in `refineMethodHandleInvokeBasic()`, but I did it there too for consistency.

jdmpapin commented:
@mpirvu, a couple of notes about the increase in JITServer messages.

First, I've been thinking about strategy 3 from my earlier comment, i.e. the strategy of doing a less precise analysis on the server and then following it with the full analysis on the client later on, and the more I think about it, the more I think we should do that. It would allow us to skip *all* of these new messages whenever it applies, so at least whenever `dontInlineUnloadableMethods` is not enabled.

We could also apply it even when `dontInlineUnloadableMethods` *is* enabled, which would allow us to completely remove these new messages. Though that would restrict inlining more than necessary, which might matter, e.g. if we want to use `dontInlineUnloadableMethods` to enable const refs on balanced.

Second, about this:

> Every (or at least nearly every) new `RetainedMethodSet` entails a scan as part of its initialization.

That's true, but thinking about it again, it's possible that a lot of these scans are trivial, e.g. when we're creating the root set for an outermost method that's defined by a class defined by a permanent loader. Depending on the workload, that might be nearly all of the sets. So I think it would help after all to only send a message for `scan`. I don't think I'd want to eliminate the remote mirrors entirely, since they can still help reduce traffic, and they also help avoid duplicating the scan implementation, but I think now that by queuing up events (strategy 1) it should be possible to eliminate all but the `scan` message.

mpirvu commented:
> it's possible that a lot of these scans are trivial,
I actually measured that and out of ~6200 mirrors, ~2800 had trivial scans that did not require looking at the fields of the class loader.

mpirvu commented:
@jdmpapin If you think that the new strategy would result in a significant reduction of messages, go for it.

jdmpapin commented:
Converted to draft while I work on improving the JITServer part.

This is still ready for review otherwise, since the changes will generally just be JITServer-related. However, I think I'll eliminate `mergeIntoParent()`, or at least change the J9 override into a no-op. I don't think there's all that much benefit to having it, and it makes it hard to reason that the server's restricted analysis and the client's full analysis will stand in a particular direct correspondence. Without `mergeIntoParent()`, the two analyses will correspond pretty straightforwardly because the contents of a set representing a call site or call target will be determined by the call path leading to it, which will be the same in both analyses.

jdmpapin commented:
Actually, I'm going to make some larger changes to the `RetainedMethodSet` API. Essentially, `attestLinkedCalleeWillRemainLoaded()` technically interferes with the following property:

> the contents of a set representing a call site or call target will be determined by the call path leading to it

I was almost but not quite able to convince myself that the discrepancies would be irrelevant. However, even if I were confident about that, the reasoning would rely on the fact that the current implementation only ever attests `LambdaForm`-generated call site adapter methods.

So I'll be replacing that with a version that creates a child set if necessary, limiting the effect of the attestation to the call site where it happens. This way, at each point in the inlining analysis, the contents of a set will be independent of the relative order in which neighbouring subtrees are built. I expect that essentially all of the benefit of these attestations is currently at the call sites where they happen anyway. They only affected other parts of the tree because originally it was easy for them to do so, and so I thought they might as well.

Because I'm potentially introducing an extra layer of `RetainedMethodSet` at a call site, I'm going to make `keepalive()` and `bond()` accumulate directly into the root set rather than into the parent, so that they don't have to propagate up the tree one level at a time. This should be a conceptual simplification because `keepalive()` and `bond()` will no longer need to be called in strictly bottom-up order.

jdmpapin commented:
Rebased to fix the conflicts, which were trivial

jdmpapin commented:
Reordered `#include`s to keep them in alphabetical order after conflict resolution

jdmpapin commented:
Pushed updates in a separate commit. I intend to squash them later.

Testing these updates exposed the problems fixed by #21838, #21925, and eclipse-omr/omr#7731.

This also now depends on #21928 so that all of the permanent loaders will be known on the server.

The changes are ready to be reviewed again, but this PR will remain in draft state until #21925 and #21928 are merged.

jdmpapin commented:
> We could also apply [strategy 3] even when `dontInlineUnloadableMethods` _is_ enabled, which would allow us to completely remove these new messages. Though that would restrict inlining more than necessary, which might matter, e.g. if we want to use `dontInlineUnloadableMethods` to enable const refs on balanced.

I did decide to always do the less precise analysis on the server and remove the new messages. The plan for const refs has changed to be much simpler on the GC side, and it will work in every GC mode without us needing to impose `dontInlineUnloadableMethods`.

jdmpapin commented:
Rebased to fix the `CommunicationStream::MINOR_NUMBER` conflict from #21928. Moving out of draft status now that all prerequisites have been merged.

jdmpapin commented:
#21928 needs to be reverted due to https://github.com/eclipse-openj9/openj9/issues/21979#issuecomment-2922622988, so I've made this a draft again :disappointed: while I work out how to deal with this kind of mutation.

But this time I'm confident that the changes that will be required in the JIT will be minimal, relating just to the way we get the permanent loaders to initialize `RetainedMethodSet` (and maybe a small tweak to deal with permanent loaders in `scan()`, if necessary). So from the JIT side this is still good to review.

dsouzai reviewed (commented):
AOT code looks ok to me.

jdmpapin commented:
Rebased and added a second update commit to make use of #22100 (and to make a small fix in the JITServer part)

The first commit in the series is amended to fix the `CommunicationStream::MINOR_NUMBER` conflict and to remove the VM changes that have already been merged as part of #22100

mpirvu reviewed (approved):
LGTM

jdmpapin commented:
Rebased

jdmpapin commented:
Rebased and fixed the conflict in CommunicationStream.hpp

0xdaryl commented on runtime/compiler/env/J9RetainedMethodSet.hpp:
```diff
@@ -0,0 +1,134 @@
+   InliningTable * const _inliningTable;
+   };
+
+} // namespace TR
```
namespace J9

0xdaryl commented on runtime/compiler/env/J9RetainedMethodSet.hpp:
```diff
@@ -0,0 +1,134 @@
+/*******************************************************************************
+ * Copyright IBM Corp. and others 2024
```
2025

0xdaryl reviewed (changes requested):
A well-designed, thoroughly researched in its detail, and extensively documented pull request.

Just a couple of minor comment corrections from me.

0xdaryl commented:
Jenkins test sanity.functional,sanity.openjdk all jdk8 depends eclipse/omr#7673
